rslib/src/collection/backup.rs

```diff
@@ -0,0 +1,480 @@
+        self.obsolete.push(backup);
+    }
+}
+
+impl Meta {
+    fn from_archive(archive: &mut ZipArchive<File>) -> Option<Self> {
+        archive
+            .by_name("meta")
+            .ok()
+            .and_then(|file| serde_json::from_reader(file).ok())
+    }
+}
+
+fn check_collection(col_path: &Path) -> Result<()> {
+    let col = CollectionBuilder::new(col_path).build()?;
+    col.storage
+        .db
+        .pragma_query_value(None, "integrity_check", |row| row.get::<_, String>(0))
+        .map_err(Into::into)
+        .and_then(|s| match s.as_str() {
+            "ok" => Ok(()),
+            _ => Err(AnkiError::invalid_input(format!("corrupt: {}", s))),
+        })
+}
+
+fn restore_media(
+    mut progress_fn: impl FnMut(usize) -> bool,
+    archive: &mut ZipArchive<File>,
+    media_folder: &str,
+) -> Result<()> {
+    let media_file_names = extract_media_file_names(archive).ok_or(AnkiError::NotFound)?;
+    let mut count = 0;
+
+    for (archive_file_name, file_name) in media_file_names {
+        count += 1;
+        if count % 10 == 0 && !progress_fn(count) {
+            return Err(AnkiError::Interrupted);
+        }
+
+        if let Ok(mut file) = archive.by_name(&archive_file_name) {
+            let file_path = Path::new(&media_folder).join(normalize_to_nfc(&file_name).as_ref());
+            let files_are_equal = fs::metadata(&file_path)
+                .map(|metadata| metadata.len() == file.size())
+                .unwrap_or_default();
+            if !files_are_equal {
+                let contents = get_zip_file_contents(&mut file).ok_or(AnkiError::NotFound)?;
+                fs::write(&file_path, &contents)?;
+            }
+        }
+    }
+    Ok(())
+}
+
+fn extract_media_file_names(archive: &mut ZipArchive<File>) -> Option<HashMap<String, String>> {
+    archive
+        .by_name("media")
+        .ok()
+        .and_then(|mut file| get_zip_file_contents(&mut file))
+        .and_then(|bytes| serde_json::from_slice(&bytes).ok())
+}
+
+fn collection_reader(archive: &mut ZipArchive<File>) -> Result<Box<dyn Read + '_>> {
+    match Meta::from_archive(archive).map(|meta| meta.version) {
+        Some(3) => version_3_reader(archive)
+            .map(|reader| Box::new(reader) as Box<dyn Read>)
+            .ok_or_else(|| AnkiError::db_error("", DbErrorKind::Corrupt)),
+        Some(_) => Err(AnkiError::db_error("", DbErrorKind::FileTooNew)),
+        None => legacy_reader(archive)
+            .map(|reader| Box::new(reader) as Box<dyn Read>)
+            .ok_or_else(|| AnkiError::db_error("", DbErrorKind::Corrupt)),
+    }
+}
+
+fn version_3_reader(archive: &mut ZipArchive<File>) -> Option<impl Read + '_> {
+    archive
+        .by_name("collection.anki21b")
+        .ok()
+        .and_then(|file| Decoder::new(file).ok())
+}
+
+fn legacy_reader<'a>(archive: &'a mut ZipArchive<File>) -> Option<impl Read + '_> {
+    if archive.file_names().contains(&"collection.anki21") {
```

Comment:
This seems needlessly wasteful, but I didn't get it to compile without searching first. The problem is that the mutable borrow of `archive` is not dropped at the end of an `if let` clause (instead only at the end of the block). Likewise, it's not dropped after a conditional return (https://users.rust-lang.org/t/solved-borrow-doesnt-drop-returning-this-value-requires-that/24182)

Reply:
Yeah, that's unfortunate. zip-rs is actually creating a hashmap of the files up front, but it's not exposed publicly. IIRC Anki clients will place the collection file in the zip before the media, so assuming the index is being read in order, this check should at least be able to terminate within a few filenames.

Reply:
... actually no, the order would be lost when the filenames are added to the hashmap. How about this?

```rust
fn legacy_reader(archive: &mut ZipArchive<File>) -> Option<impl Read + '_> {
    let have21 = archive.by_name("collection.anki21").is_ok();
    if have21 {
        archive.by_name("collection.anki21").ok()
    } else {
        archive.by_name("collection.anki2").ok()
    }
}
```